core/internals.h

```diff
@@ -64,31 +64,78 @@
-    if ((URI) == NULL) lwm2m_printf("[%s:%d] NULL\r\n", __func__ , __LINE__);       \
-    else                                                                            \
-    {                                                                               \
-        lwm2m_printf("[%s:%d] /%d", __func__ , __LINE__ , (URI)->objectId);         \
-        if (LWM2M_URI_IS_SET_INSTANCE(URI)) lwm2m_printf("/%d", (URI)->instanceId); \
-        if (LWM2M_URI_IS_SET_RESOURCE(URI)) lwm2m_printf("/%d", (URI)->resourceId); \
-        lwm2m_printf("\r\n");                                                       \
-    }                                                                               \
-}
+
+#ifdef _WIN32
+#define LWM2M_NEW_LINE "\r\n"
 #else
-#define LOG_URI(URI)                                                                \
-    if ((URI) == NULL) lwm2m_printf("[%s:%d] NULL\r\n", __func__ , __LINE__);       \
-    else if (!LWM2M_URI_IS_SET_OBJECT(URI)) lwm2m_printf("[%s:%d] /\r\n", __func__ , __LINE__); \
-    else if (!LWM2M_URI_IS_SET_INSTANCE(URI)) lwm2m_printf("[%s:%d] /%d\r\n", __func__ , __LINE__, (URI)->objectId); \
-    else if (!LWM2M_URI_IS_SET_RESOURCE(URI)) lwm2m_printf("[%s:%d] /%d/%d\r\n", __func__ , __LINE__, (URI)->objectId, (URI)->instanceId); \
-    else if (!LWM2M_URI_IS_SET_RESOURCE_INSTANCE(URI)) lwm2m_printf("[%s:%d] /%d/%d/%d\r\n", __func__ , __LINE__, (URI)->objectId, (URI)->instanceId, (URI)->resourceId); \
-    else lwm2m_printf("[%s:%d] /%d/%d/%d/%d\r\n", __func__ , __LINE__, (URI)->objectId, (URI)->instanceId, (URI)->resourceId, (URI)->resourceInstanceId)
+#define LWM2M_NEW_LINE "\n"
 #endif
+
+#ifndef LWM2M_LOG_MAX_MSG_TXT_SIZE
+#define LWM2M_LOG_MAX_MSG_TXT_SIZE 200
+#endif
+
+typedef enum {
+    LWM2M_LOGGING_DBG = (uint8_t)LWM2M_DBG,
+    LWM2M_LOGGING_INFO = (uint8_t)LWM2M_INFO,
+    LWM2M_LOGGING_WARN = (uint8_t)LWM2M_WARN,
+    LWM2M_LOGGING_ERR = (uint8_t)LWM2M_ERR,
+    LWM2M_LOGGING_FATAL = (uint8_t)LWM2M_FATAL
+} lwm2m_logging_level_t;
+
+/* clang-format off */
+#define STR_LOGGING_LEVEL(level) \
+((level) == LWM2M_LOGGING_DBG ? "DBG" : \
+((level) == LWM2M_LOGGING_INFO ? "INFO" : \
+((level) == LWM2M_LOGGING_WARN ? "WARN" : \
+((level) == LWM2M_LOGGING_ERR ? "ERR" : \
+((level) == LWM2M_LOGGING_FATAL ? "FATAL" : \
+"unknown")))))
+/* clang-format on */
+
+/** Format the message part of a log entry. This function is not thread save.
+ * This function should not be called directly. It's used internally in the logging system. */
+char *lwm2m_log_fmt_message(const char *fmt, ...);
+
+/** The default log handler for an log entry. To define a custom log handler define `LWM2M_LOG_CUSTOM_HANDLER` and
+ * implement a function with this signature.
+ * This function should not be called directly. Use the logging macros instead. */
+void lwm2m_log_handler(lwm2m_logging_level_t level, const char *const msg, const char *const func, const int line,
```

Comment:
How about putting this declaration in the "public" header? This might be needed by a user of the library and would require to include the internals header.